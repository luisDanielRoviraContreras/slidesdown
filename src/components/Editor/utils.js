export const getCurrentLineNumber = (value, cursorPosition) =>
  value.substr(0, cursorPosition).split('\n').length
