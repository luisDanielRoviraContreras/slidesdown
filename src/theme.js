export default {
  colors: {
    border: '#eee',
    magenta: '#c86dd7',
    purple: '#3023ae'
  },
  fonts: {
    default: 'Montserrat',
    monospace: 'Inconsolata'
  }
}
