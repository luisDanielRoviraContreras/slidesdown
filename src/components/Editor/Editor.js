import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { getCurrentLineNumber, getSlidesFirstLines } from './utils'

const StyledTextarea = styled.textarea`
  border: 0;
  color: ${props => props.theme.colors.lightText};
  font-family: ${props => props.theme.fonts.monospace};
  font-size: 16px;
  height: 100%;
  left: 0;
  outline: none;
  padding: 1rem;
  position: absolute;
  top: 0;
  resize: none;
  width: 100%;
  -webkit-overflow-scrolling: touch;
`

class Editor extends Component {
  static propTypes = {
    onChange: PropTypes.func.isRequired,
    onCursorPositionChange: PropTypes.func,
    onSlideChange: PropTypes.func,
    value: PropTypes.string
  }

  getCurrentCursorPosition = () => this.editor.selectionStart

  getCurrentSlide = e => {
    const { onSlideChange, value } = this.props

    const currentLineNumber = getCurrentLineNumber(
      this.editor.value,
      this.getCurrentCursorPosition()
    )

    const slides = getSlidesFirstLines(value)

    onSlideChange &&
      onSlideChange(
        slides.reverse().find(slide => currentLineNumber > slide.firstLine)
          .slide
      )
  }

  handleCursorPositionChange = e => {
    const { onCursorPositionChange } = this.props

    onCursorPositionChange &&
      onCursorPositionChange(this.getCurrentCursorPosition())
  }

  render() {
    const { onChange, value } = this.props

    return (
      <StyledTextarea
        innerRef={ref => (this.editor = ref)}
        onChange={onChange}
        onClick={this.getCurrentSlide}
        onKeyUp={this.getCurrentSlide}
        placeholder="Write markdown here"
        value={value}
      />
    )
  }
}

export default Editor
