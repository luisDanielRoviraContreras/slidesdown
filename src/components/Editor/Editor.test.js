import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { ThemeProvider } from 'styled-components'

import theme from '../../theme'
import Editor from '../Editor'

it('renders correctly', () => {
  const onChangeMock = jest.fn()

  const tree = renderer
    .create(
      <ThemeProvider theme={theme}>
        <Editor onChange={onChangeMock} value="Test" />
      </ThemeProvider>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
