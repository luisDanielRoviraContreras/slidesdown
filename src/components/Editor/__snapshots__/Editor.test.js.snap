// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  border: 0;
  color: #646468;
  font-family: Inconsolata;
  font-size: 16px;
  height: 100%;
  left: 0;
  outline: none;
  padding: 1rem;
  position: absolute;
  top: 0;
  resize: none;
  width: 100%;
  -webkit-overflow-scrolling: touch;
}

<textarea
  className="c0"
  onChange={[MockFunction]}
  onClick={[Function]}
  onKeyUp={[Function]}
  placeholder="Write markdown here"
  value="Test"
/>
`;
